### PR TITLE
add perception encoder to inference_experimental

### PR DESCRIPTION
**Pull request by hansent**
# Description
First shot at adding perception encoder to inference_experimental 

## Type of change
new model in experimental package

## How has this change been tested, please provide a testcase or example of how you tested the change?
locally with weight downloaded in tmp dir and pointing test to load it from there

## Any specific deployment considerations
n/a

## Docs
n/a

### CODE REVIEW

File: inference_experimental/inference_exp/models/perception_encoder/perception_encoder.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
+import os
+from typing import List, Union
+
+import numpy as np
+import torch
+from PIL import Image
+
+import inference_exp.models.perception_encoder.vision_encoder.pe as pe
```
Comment (hansent) [inline inference_experimental/inference_exp/models/perception_encoder/perception_encoder.py]: wondering if these could/should all be part of model package now?

File: inference_experimental/inference_exp/models/perception_encoder/perception_encoder.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        self.preprocessor = preprocessor
+        self.tokenizer = tokenizer
+        self.device = device
+
+    @classmethod
+    def from_pretrained(
+        cls, model_name_or_path: str, device: str = "cpu", **kwargs
+    ) -> "PerceptionEncoder":
+        #here model name came from path before, which maybe doesn't match directly with how our registry works
+        # instead should this be adopted to read config file that is served as part of model package?
```
Comment (hansent) [inline inference_experimental/inference_exp/models/perception_encoder/perception_encoder.py]: is convention that we provide full name to weights file?  in this case it needs some additional info (to load propper config based on https://github.com/roboflow/inference/pull/1399/files#diff-1f145eebc7319d56dbf1e831f4f662f456fd0468af00eedf2b39a095927e165bR89)

File: inference_experimental/inference_exp/models/perception_encoder/perception_encoder.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        elif isinstance(image, torch.Tensor):
+            pil_image = Image.fromarray(image.cpu().numpy())
+        else:
+            pil_image = image
+        preprocessed_image = self.preprocessor(pil_image)
+        return preprocessed_image.unsqueeze(0)
+
+    def embed_images(
+        self,
+        images: Union[torch.Tensor, List[torch.Tensor], np.ndarray, List[np.ndarray], List[Image.Image]],
```
Comment (hansent) [inline inference_experimental/inference_exp/models/perception_encoder/perception_encoder.py]: without Iamge.Image?

File: inference_experimental/inference_exp/models/perception_encoder/perception_encoder.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        **kwargs,
+    ) -> torch.Tensor:
+        if isinstance(texts, list):
+            texts_to_embed = texts
+        else:
+            texts_to_embed = [texts]
+
+        #results = []
+        # The original implementation had batching here based on CLIP_MAX_BATCH_SIZE, but not entirely sure how to handle that with Tensor output
+        # I will leave it out for now, see https://github.com/roboflow/inference/blob/main/inference/models/perception_encoder/perception_encoder.py#L227
```
Comment (hansent) [inline inference_experimental/inference_exp/models/perception_encoder/perception_encoder.py]: existing implementation in inferences uses create_batches utility to split into max size batches, but wasnt sure how to maintain return type

File: inference_experimental/tests/integration_tests/models/test_perception_encoder.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+from inference_exp.models.perception_encoder.perception_encoder import (
+    PerceptionEncoder,
+)
+
+
+def test_perception_encoder_text_embedding():
+
+    #TODO: this is a temporary path, should be replaced with a proper path / registry integration
+    model = PerceptionEncoder.from_pretrained(
+        "/tmp/cache/perception_encoder/PE-Core-B16-224"
```
Comment (hansent) [inline inference_experimental/tests/integration_tests/models/test_perception_encoder.py]: just have this locally t the moment so I could do sanity check (to download can use PE in workflow block with existing inference) until added to registry

File: inference_experimental/inference_exp/models/perception_encoder/perception_encoder.py (hunk 2 above)
Comment (PawelPeczek-Roboflow) [reply inference_experimental/inference_exp/models/perception_encoder/perception_encoder.py]: we should not rely on that, backend will associate hashed ID for model package and then you would have cache dir in format `<prefix>/<model_id>/<model_package_id>`

Comment (PawelPeczek-Roboflow) [reply inference_experimental/inference_exp/models/perception_encoder/perception_encoder.py]: I would simply use the util for indexing the contents of the directory to get mapping from file-name to path.
By fixing file names we set an interface between code registering model and inference
Mentioned util `get_model_package_contents(...)`

File: inference_experimental/inference_exp/models/perception_encoder/perception_encoder.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        device: str,
+    ):
+        self.model = model
+        self.preprocessor = preprocessor
+        self.tokenizer = tokenizer
+        self.device = device
+
+    @classmethod
+    def from_pretrained(
+        cls, model_name_or_path: str, device: str = "cpu", **kwargs
```
Comment (PawelPeczek-Roboflow) [inline inference_experimental/inference_exp/models/perception_encoder/perception_encoder.py]: there is default device variable in `constants.py`

File: inference_experimental/inference_exp/models/perception_encoder/perception_encoder.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    ) -> "PerceptionEncoder":
+        #here model name came from path before, which maybe doesn't match directly with how our registry works
+        # instead should this be adopted to read config file that is served as part of model package?
+        model_config = model_name_or_path.split("/")[-1]
```
Comment (PawelPeczek-Roboflow) [inline inference_experimental/inference_exp/models/perception_encoder/perception_encoder.py]: if additional metadata needed, should be assumed that backend provide config file to be parsed rather than decoding paths

File: inference_experimental/inference_exp/models/perception_encoder/perception_encoder.py (hunk 3 above)
Comment (PawelPeczek-Roboflow) [reply inference_experimental/inference_exp/models/perception_encoder/perception_encoder.py]: without - we have 2 in grid already, adding PIL means we must implement all pre-processing taking that as an alternative - wanted to stick to torch only but for jetson performance reasons also allowed numpy

File: inference_experimental/inference_exp/models/perception_encoder/perception_encoder.py (hunk 4 above)
Comment (PawelPeczek-Roboflow) [reply inference_experimental/inference_exp/models/perception_encoder/perception_encoder.py]: we can do the same, then probably `torch.cat([tensor_1, ..., tensor_n], dim=0)` will be needed

**Review by PawelPeczek-Roboflow — COMMENTED** [review]
looks good up to minor changes

File: inference_experimental/inference_exp/models/perception_encoder/perception_encoder_pytorch.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+            image = torch.from_numpy(image).permute(2, 0, 1)
+            # BGR -> RGB
+            image = image[[2, 1, 0], :, :]
+
+        if not isinstance(image, torch.Tensor):
+            raise TypeError(
+                "Unsupported image type, must be np.ndarray or torch.Tensor"
+            )
+
+        # The original ToTensor() transform also scaled images from [0, 255] to [0, 1].
```
Comment (PawelPeczek-Roboflow) [inline inference_experimental/inference_exp/models/perception_encoder/perception_encoder_pytorch.py]: 👍

File: inference_experimental/inference_exp/models/perception_encoder/perception_encoder_pytorch.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+            model_package_dir=model_name_or_path,
+            elements=["config.json", "model.pt"],
+        )
+
+        model_config_file = model_package_content["config.json"]
+        model_weights_file = model_package_content["model.pt"]
+        with open(model_config_file) as f:
+            config = json.load(f)
+
+        print(
```
Comment (PawelPeczek-Roboflow) [inline inference_experimental/inference_exp/models/perception_encoder/perception_encoder_pytorch.py]: [1] debug print

File: inference_experimental/inference_exp/models/perception_encoder/perception_encoder_pytorch.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+        # checkpoint_path = os.path.join(model_name_or_path, "model.pt")
+
+        model_package_content = get_model_package_contents(
+            model_package_dir=model_name_or_path,
+            elements=["config.json", "model.pt"],
+        )
+
+        model_config_file = model_package_content["config.json"]
+        model_weights_file = model_package_content["model.pt"]
+        with open(model_config_file) as f:
```
Comment (PawelPeczek-Roboflow) [inline inference_experimental/inference_exp/models/perception_encoder/perception_encoder_pytorch.py]: there is an util if you want - in general I believe it would be super cool if all models follow the same pattern in raising  errors - here I mean for instance raising `CorruptedModelPackageError` - example from YoloV8:
```python
    @classmethod
    def from_pretrained(
        cls,
        model_name_or_path: str,
        device: torch.device = DEFAULT_DEVICE,
        engine_host_code_allowed: bool = False,
        **kwargs,
    ) -> "YOLOv8ForInstanceSegmentationTRT":
        if device.type != "cuda":
            raise ModelRuntimeError(
                message=f"TRT engine only runs on CUDA device - {device} device detected.",
                help_url="https://todo",
            )
       [...]
```

Comment (PawelPeczek-Roboflow) [reply inference_experimental/inference_exp/models/perception_encoder/perception_encoder_pytorch.py]: I mean - there is full set of library errors prepared, aiming to also provide meaningful error messages (and in the future direct links to docs with potential solution)

File: inference_experimental/inference_exp/models/perception_encoder/perception_encoder_pytorch.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+                "Unsupported image type, must be np.ndarray or torch.Tensor"
+            )
+
+        # The original ToTensor() transform also scaled images from [0, 255] to [0, 1].
+        # We need to replicate that behavior.
+        if image.dtype == torch.uint8:
+            image = image.to(torch.float32) / 255.0
+
+        preprocessed_image = preprocessor(image)
+        return preprocessed_image.unsqueeze(0)
```
Comment (PawelPeczek-Roboflow) [inline inference_experimental/inference_exp/models/perception_encoder/perception_encoder_pytorch.py]: hmm - ok this strikes to my attention as potential issue - since we `unsqueeze` it looks as if we are expanding towrads `N` dimension to have `(N, C, H, W)`, right?

then we have preprocessor applied in the following way:
```python
    if isinstance(images, list):
            imgs = [self.preprocessor(i) for i in images]
            img_in = torch.cat(imgs, dim=0).to(self.device)
        else:
            img_in = self.preprocessor(images).to(self.device)
```

input may be:
* single `np.ndarray` - handled
* list of `np.ndarray` - handled
* `torch.Tensor` of shape `(C, H, W)` - handled
* list of `torch.Tensor` of shape `(C, H, W)` - handled
* `torch.Tensor` of shape `(N, C, H, W)` ⚠️  - the only problem being as I got the following working:

```
image = torch.zeros((10, 3, 1080, 1920), dtype=torch.float)
transforms = get_tensor_image_transform(640)
result = transforms(image)
result.shape
>>> torch.Size([10, 3, 640, 640])
```

File: inference_experimental/inference_exp/models/perception_encoder/perception_encoder_pytorch.py [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+
+        return cls(
+            model=model,
+            device=device,
+        )
+
+    def embed_images(
+        self,
+        images: Union[torch.Tensor, List[torch.Tensor], np.ndarray, List[np.ndarray]],
+        **kwargs,
```
Comment (PawelPeczek-Roboflow) [inline inference_experimental/inference_exp/models/perception_encoder/perception_encoder_pytorch.py:160]: I believe it would be good if we keep this input parameter dictating the input colour format with default `None` and meaning of the default anchored to the type of data.

Then you would need to slightly alter:

```python
def _preprocess_image(image: Union[np.ndarray, torch.Tensor]) -> torch.Tensor:
```

to look like

```python
def _preprocess_image(
    image: Union[np.ndarray, torch.Tensor], 
    input_color_format: Optional[ColorFormat] = None,
) -> torch.Tensor:
```

File: inference_experimental/tests/integration_tests/models/test_perception_encoder.py [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+
+        preprocessed_image = original_pil_transform(pil_image)
+        return preprocessed_image.unsqueeze(0)
+
+    return _pil_preprocessor
+
+
+@pytest.mark.parametrize("model_size", [224, 336, 448])
+@pytest.mark.parametrize("image_shape", [(224, 224), (336, 448), (640, 640)])
+def test_preprocessing_consistency(model_size, image_shape):
```
Comment (PawelPeczek-Roboflow) [inline inference_experimental/tests/integration_tests/models/test_perception_encoder.py:38]: looks great - the only think I would change is to have separate test for numpy and tensor inputs (and when support for (N, C, H, W) tensors added - test covering that to iterative application against list of PIL)

File: inference_experimental/tests/integration_tests/models/test_perception_encoder.py [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+
+    # THEN
+    assert torch.allclose(numpy_tensor_output, numpy_pil_output, atol=1e-2)
+    assert torch.allclose(tensor_tensor_output, tensor_pil_output, atol=1e-2)
+    # Also ensure that processing a numpy array vs a tensor gives the same result
+    assert torch.allclose(numpy_tensor_output, tensor_tensor_output, atol=1e-2)
+
+
+@pytest.mark.e2e_model_inference
+def test_perception_encoder_text_embedding():
```
Comment (PawelPeczek-Roboflow) [inline inference_experimental/tests/integration_tests/models/test_perception_encoder.py]: [2] I would move this into separate suite - maybe `tests/integration_tests/e2e` (or sth better) for clarity

File: inference_experimental/inference_exp/models/perception_encoder/perception_encoder_pytorch.py (hunk 12 above)
Comment (hansent) [reply inference_experimental/inference_exp/models/perception_encoder/perception_encoder_pytorch.py:160]: as param on embed_image and then pass through to preprocessing?  on other models base interfaces we have very specific interface for pre_process, forward_pass and infer ensuring the forward gets called with pre_processed input images.   Does it make sense to do this for TextImageEmbeddingModel also..or maybe it doesnt matter? The actual implementations on the other classes expose the pre_processing params via named keyword args to some extent, but I guess it also gets hidden for infer taking just **kwargs.  Wondering if its useful for the function signatures to communicate model specific pre-processing params (as opposed to *kwargs) or if that doesn't make sense

File: inference_experimental/inference_exp/models/perception_encoder/perception_encoder_pytorch.py (hunk 10 above)
Comment (hansent) [reply inference_experimental/inference_exp/models/perception_encoder/perception_encoder_pytorch.py]: :+1:  

added config loading/parsing in the model itself since its sepcific format, but using the common errors:
```python
class PerceptionEncoderConfig(BaseModel):
    vision_encoder_config: str


def load_config(config_path: str) -> PerceptionEncoderConfig:
    config_data = {}
    try:
        with open(config_path) as f:
            config_data = json.load(f)
    except (IOError, json.JSONDecodeError) as e:
        raise CorruptedModelPackageError(
            message=f"Could not load or parse perception encoder model package config file: {config_path}. Details: {e}",
            help_url="https://todo",
        ) from e
    try:
        config = PerceptionEncoderConfig.model_validate(config_data)
        return config
    except ValidationError as e:
        raise CorruptedModelPackageError(
            f"Failed validate perception encoder model package config file: {config_path}. Details: {e}"
        ) from e
```

File: inference_experimental/inference_exp/models/perception_encoder/perception_encoder_pytorch.py (hunk 11 above)
Comment (hansent) [reply inference_experimental/inference_exp/models/perception_encoder/perception_encoder_pytorch.py]: changed preproc to handle the list/batch/dimensions itself and adjusted tests to cover all cases and ended up optimizing a bit more to do the float conversion on batched tensor at the end.  Had to split the transform defintion into two so we can resize first in case images have different dimensions
```python
def create_preprocessor(image_size: int) -> Callable:
    resize_transform = create_image_resize_transform(image_size)
    normalize_transform = create_image_normalize_transform()

    def _preprocess(
        images: Union[torch.Tensor, List[torch.Tensor], np.ndarray, List[np.ndarray]]
    ) -> torch.Tensor:
        def _to_tensor(image: Union[np.ndarray, torch.Tensor]) -> torch.Tensor:
            if isinstance(image, np.ndarray):
                # HWC -> CHW
                image = torch.from_numpy(image).permute(2, 0, 1)
                # BGR -> RGB
                image = image[[2, 1, 0], :, :]
            return image

        if isinstance(images, list):
            # Resize each image individually, then stack to a batch
            resized_images = [resize_transform(_to_tensor(img)) for img in images]
            tensor_batch = torch.stack(resized_images, dim=0)
        else:
            # Handle single image or pre-batched tensor
            tensor_batch = resize_transform(_to_tensor(images))

        # Ensure there is a batch dimension for single images
        if tensor_batch.ndim == 3:
            tensor_batch = tensor_batch.unsqueeze(0)

        # Perform dtype conversion and normalization on the whole batch for efficiency
        if tensor_batch.dtype == torch.uint8:
            tensor_batch = tensor_batch.to(torch.float32) / 255.0

        transformed_batch = normalize_transform(tensor_batch)
        return transformed_batch

    return _preprocess
```

Comment (hansent) [reply inference_experimental/inference_exp/models/perception_encoder/perception_encoder_pytorch.py]: ended up optimizing a bit more to do the float conversion on batched tensor at the end.  Had to split the transform defintion into two so we can resize first in case images have different dimensions

File: inference_experimental/tests/integration_tests/models/test_perception_encoder.py (hunk 13 above)
Comment (hansent) [reply inference_experimental/tests/integration_tests/models/test_perception_encoder.py:38]: done

File: inference_experimental/tests/integration_tests/models/test_perception_encoder.py (hunk 14 above)
Comment (hansent) [reply inference_experimental/tests/integration_tests/models/test_perception_encoder.py]: done

File: inference_experimental/inference_exp/models/perception_encoder/perception_encoder_pytorch.py (hunk 12 above)
Comment (hansent) [reply inference_experimental/inference_exp/models/perception_encoder/perception_encoder_pytorch.py:160]: This is only thing left I think, lets sync on it not exactly sure what best way to pass it through/expose param to user

Comment (hansent) [reply inference_experimental/inference_exp/models/perception_encoder/perception_encoder_pytorch.py:160]: added 👍

**Review by PawelPeczek-Roboflow — APPROVED** [review]
